Source/WebCore/Modules/mediastream/VideoTrackGenerator.h

```diff
@@ -37,16 +40,63 @@ class WritableStream;
 class VideoTrackGenerator : public RefCounted<VideoTrackGenerator> {
     WTF_MAKE_ISO_ALLOCATED(VideoTrackGenerator);
 public:
-    static Ref<VideoTrackGenerator> create(ScriptExecutionContext& context) { return adoptRef(*new VideoTrackGenerator(context)); }
+    static ExceptionOr<Ref<VideoTrackGenerator>> create(ScriptExecutionContext&);
     ~VideoTrackGenerator();
 
-    void setMuted(bool);
-    bool muted() const { return m_muted; }
+    void setMuted(ScriptExecutionContext&, bool);
+    bool muted(ScriptExecutionContext&) const { return m_muted; }
+
+    Ref<WritableStream> writable();
+    Ref<MediaStreamTrack> track();
 
 private:
-    explicit VideoTrackGenerator(ScriptExecutionContext&);
+    class Sink;
+    VideoTrackGenerator(Ref<Sink>&&, Ref<WritableStream>&&, Ref<MediaStreamTrack>&&);
+
+    class Source final : public RealtimeMediaSource, public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Source, WTF::DestructionThread::MainRunLoop> {
+    public:
+        static Ref<Source> create() { return adoptRef(*new Source()); }
+
+        void ref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Source, WTF::DestructionThread::MainRunLoop>::ref(); }
+        void deref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Source, WTF::DestructionThread::MainRunLoop>::deref(); }
+        ThreadSafeWeakPtrControlBlock& controlBlock() const final { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Source, WTF::DestructionThread::MainRunLoop>::controlBlock(); }
+
+        void writeVideoFrame(VideoFrame&, VideoFrameTimeMetadata);
+
+    private:
+        Source();
+
+        const RealtimeMediaSourceCapabilities& capabilities() final { return m_capabilities; }
+        const RealtimeMediaSourceSettings& settings() final { return m_settings; }
```

Comment:
It looks like these aren't used, are they needed?

Reply:
They are needed as the base class marks them as virtual pure.